### PR TITLE
config_consistency: Enable tests for the python and ruby tracers

### PR DESCRIPTION
**Pull request by mabdinur**
## Motivation

- Enables all python and ruby config consistency tests that are XPassing.

### Changes to server.rb

- Updates the `start_span` endpoint in the ruby parametric app to avoid setting service, resource, and type arguments to an empty string by default. The default values for these arguments MUST be nil.
- Ensures the `find_span` helper in the ruby parametric app can retrieve spans that are not active. There is not need to only allow parametric endpoints to operate on active spans. This change allows us to test scenarios when the tracing api is used AND DD_TRACE_ENABLED=False.

### Minor changes (not required)

- Refactors tests in tests/parametric/test_config_consistency.py to improve readability

## Workflow

1. ⚠️ Create your PR as draft ⚠️
2. Work on you PR until the CI passes (if something not related to your task is failing, you can ignore it)
3. Mark it as ready for review
    * Test logic is modified? -> Get a review from RFC owner. We're working on refining the `codeowners` file quickly.
    * Framework is modified, or non obvious usage of it -> get a review from [R&P team](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X)

:rocket: Once your PR is reviewed, you can merge it!

🛟 [#apm-shared-testing](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X) 🛟

## Reviewer checklist

* [ ] If PR title starts with `[<language>]`, double-check that only `<language>` is impacted by the change
* [ ] No system-tests internal is modified. Otherwise, I have the approval from [R&P team](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X)
* [ ] CI is green, or failing jobs are not related to this change (and you are 100% sure about this statement)
* [ ] A docker base image is modified?
    * [ ] the relevant `build-XXX-image` label is present
* [ ] A scenario is added (or removed)?
    * [ ] Get a review from [R&P team](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X)

### CODE REVIEW

File: manifests/nodejs.yml [hunk 1]
```diff
@@ -574,7 +574,7 @@ tests/:
     Test_Config_HttpClientErrorStatuses_FeatureFlagCustom: missing_feature
     Test_Config_HttpServerErrorStatuses_Default: missing_feature
     Test_Config_HttpServerErrorStatuses_FeatureFlagCustom: missing_feature
-    Test_Config_UnifiedServiceTagging_CustomService: missing_feature
+    Test_Config_UnifiedServiceTagging_CustomService: *ref_3_0_0 (feature should be supported in all recent versions)
```
Comment (mabdinur) [inline manifests/nodejs.yml:577]:
```suggestion
    # feature should be supported in all recent versions
    Test_Config_UnifiedServiceTagging_CustomService: *ref_3_0_0
```

File: manifests/php.yml [hunk 2]
```diff
@@ -327,7 +327,7 @@ tests/:
     Test_Config_HttpClientErrorStatuses_FeatureFlagCustom: missing_feature
     Test_Config_HttpServerErrorStatuses_Default: v1.3.0 # Unknown initial version
     Test_Config_HttpServerErrorStatuses_FeatureFlagCustom: missing_feature
-    Test_Config_UnifiedServiceTagging_CustomService: missing_feature
+    Test_Config_UnifiedServiceTagging_CustomService: *ref_3_0_0 (feature should be supported in all recent versions)
```
Comment (mabdinur) [inline manifests/php.yml:330]:
```suggestion
    Test_Config_UnifiedServiceTagging_CustomService: v1.0.0 (feature should be supported in all recent versions)
```

File: manifests/cpp.yml [hunk 3]
```diff
@@ -179,7 +179,7 @@ tests/:
     Test_Config_HttpClientErrorStatuses_FeatureFlagCustom: missing_feature
     Test_Config_HttpServerErrorStatuses_Default: missing_feature
     Test_Config_HttpServerErrorStatuses_FeatureFlagCustom: missing_feature
-    Test_Config_UnifiedServiceTagging_CustomService: missing_feature
+    Test_Config_UnifiedServiceTagging_CustomService: v0.2.2
```
Comment (mabdinur) [inline manifests/cpp.yml:182]:
```suggestion
    Test_Config_UnifiedServiceTagging_CustomService: missing_feature
```

File: manifests/dotnet.yml [hunk 4]
```diff
@@ -377,7 +377,7 @@ tests/:
     Test_Config_HttpClientErrorStatuses_FeatureFlagCustom: missing_feature
     Test_Config_HttpServerErrorStatuses_Default: missing_feature
     Test_Config_HttpServerErrorStatuses_FeatureFlagCustom: missing_feature
-    Test_Config_UnifiedServiceTagging_CustomService: missing_feature
+    Test_Config_UnifiedServiceTagging_CustomService: v2.0.0
```
Comment (mabdinur) [inline manifests/dotnet.yml:380]:
```suggestion
    Test_Config_UnifiedServiceTagging_CustomService: missing_feature
```

File: manifests/golang.yml [hunk 5]
```diff
@@ -494,7 +494,7 @@ tests/:
     Test_Config_HttpClientErrorStatuses_FeatureFlagCustom: missing_feature
     Test_Config_HttpServerErrorStatuses_Default: missing_feature
     Test_Config_HttpServerErrorStatuses_FeatureFlagCustom: missing_feature
-    Test_Config_UnifiedServiceTagging_CustomService: missing_feature
+    Test_Config_UnifiedServiceTagging_CustomService: v2.0.0
```
Comment (mabdinur) [inline manifests/golang.yml:497]:
```suggestion
    Test_Config_UnifiedServiceTagging_CustomService: missing_feature
```

File: manifests/java.yml [hunk 6]
```diff
@@ -1291,7 +1291,7 @@ tests/:
     Test_Config_HttpClientErrorStatuses_FeatureFlagCustom: missing_feature
     Test_Config_HttpServerErrorStatuses_Default: missing_feature
     Test_Config_HttpServerErrorStatuses_FeatureFlagCustom: missing_feature
-    Test_Config_UnifiedServiceTagging_CustomService: missing_feature
+    Test_Config_UnifiedServiceTagging_CustomService: v1.0.0
```
Comment (mabdinur) [inline manifests/java.yml:1294]:
```suggestion
    Test_Config_UnifiedServiceTagging_CustomService: missing_feature
```

File: manifests/nodejs.yml [hunk 7]
```diff
@@ -574,7 +574,8 @@ tests/:
     Test_Config_HttpClientErrorStatuses_FeatureFlagCustom: missing_feature
     Test_Config_HttpServerErrorStatuses_Default: missing_feature
     Test_Config_HttpServerErrorStatuses_FeatureFlagCustom: missing_feature
-    Test_Config_UnifiedServiceTagging_CustomService: missing_feature
+    # feature should be supported in all recent versions
+    Test_Config_UnifiedServiceTagging_CustomService: *ref_3_0_0
```
Comment (mabdinur) [inline manifests/nodejs.yml:578]:
```suggestion
    Test_Config_UnifiedServiceTagging_CustomService: missing_feature
```

File: manifests/php.yml [hunk 8]
```diff
@@ -327,7 +327,7 @@ tests/:
     Test_Config_HttpClientErrorStatuses_FeatureFlagCustom: missing_feature
     Test_Config_HttpServerErrorStatuses_Default: v1.3.0 # Unknown initial version
     Test_Config_HttpServerErrorStatuses_FeatureFlagCustom: missing_feature
-    Test_Config_UnifiedServiceTagging_CustomService: missing_feature
+    Test_Config_UnifiedServiceTagging_CustomService: v1.0.0
```
Comment (mabdinur) [inline manifests/php.yml:330]:
```suggestion
    Test_Config_UnifiedServiceTagging_CustomService: missing_feature
```

File: tests/parametric/test_config_consistency.py [hunk 9]
```diff
@@ -26,11 +26,10 @@ class Test_Config_TraceEnabled:
     def test_tracing_enabled(self, library_env, test_agent, test_library):
         assert library_env.get("DD_TRACE_ENABLED", "true") == "true"
         with test_library:
-            with test_library.start_span("allowed"):
+            with test_library.start_span("allowed") as s:
```
Comment (mabdinur) [inline tests/parametric/test_config_consistency.py:29]:
```suggestion
            with test_library.start_span("allowed"):
```

File: tests/parametric/test_config_consistency.py [hunk 10]
```diff
@@ -99,12 +97,12 @@ def test_specific_version(self, library_env, test_agent, test_library):
         assert len(traces) == 2
 
         span1 = find_span_in_traces(traces, s1.trace_id, s1.span_id)
-        assert span1["service"] == "version_test"
-        assert span1["meta"]["version"] == "5.2.0"
+        assert span1["service"] == "version_test", f"{span1}"
+        assert span1["meta"]["version"] == "5.2.0", f"{span1}"
```
Comment (mabdinur) [inline tests/parametric/test_config_consistency.py:101]:
```suggestion
        assert span1["service"] == "version_test"
        assert span1["meta"]["version"] == "5.2.0"
```

**Comment by mabdinur** [comment]
Failing ruby tests are unreleated

**Review by khanayan123 — APPROVED** [review]

File: manifests/python.yml [hunk 11]
```diff
@@ -772,16 +772,18 @@ tests/:
...
     Test_Config_ClientIPHeader_Precedence: missing_feature (all headers listed in the RFC are not supported)
-    Test_Config_ClientTagQueryString_Configured: missing_feature (supports DD_HTPP_CLIENT_TAGS_QUERY_STRING instead)
+    Test_Config_ClientTagQueryString_Configured: v2.14.0.dev
     Test_Config_ClientTagQueryString_Empty: v2.12.0
     Test_Config_HttpClientErrorStatuses_Default: missing_feature
     Test_Config_HttpClientErrorStatuses_FeatureFlagCustom: missing_feature
-    Test_Config_HttpServerErrorStatuses_Default: missing_feature
+    Test_Config_HttpServerErrorStatuses_Default: v2.12.0
     Test_Config_HttpServerErrorStatuses_FeatureFlagCustom: missing_feature
     Test_Config_ObfuscationQueryStringRegexp_Configured: missing_feature
```
Comment (mabdinur) [inline manifests/python.yml:781]:
```suggestion
    Test_Config_ObfuscationQueryStringRegexp_Configured: v2.0.0
```

**Review by cbeauchesne — APPROVED** [review]

File: manifests/python.yml [hunk 12]
```diff
@@ -773,18 +773,20 @@ tests/:
...
-    Test_Config_HttpServerErrorStatuses_FeatureFlagCustom: missing_feature
-    Test_Config_IntegrationEnabled_False: missing_feature
-    Test_Config_IntegrationEnabled_True: missing_feature
-    Test_Config_ObfuscationQueryStringRegexp_Configured: missing_feature
+    Test_Config_HttpServerErrorStatuses_Default: v2.12.0
+    Test_Config_HttpServerErrorStatuses_FeatureFlagCustom: v2.15.0
+    Test_Config_IntegrationEnabled_False: v2.0.0
+    Test_Config_IntegrationEnabled_True: v2.0.0
+    Test_Config_ObfuscationQueryStringRegexp_Configured: v2.0.0
     Test_Config_ObfuscationQueryStringRegexp_Empty: missing_feature
```
Comment (mabdinur) [inline manifests/python.yml:785]:
```suggestion
    Test_Config_ObfuscationQueryStringRegexp_Empty: v2.0.0
```

File: manifests/python.yml [hunk 13]
```diff
@@ -786,19 +786,19 @@ tests/:
       Test_Miscs: missing_feature
   test_config_consistency.py:
     Test_Config_ClientIPHeader_Configured: v2.12.0
-    Test_Config_ClientIPHeader_Precedence: missing_feature (all headers listed in the RFC are not supported)
-    Test_Config_ClientTagQueryString_Configured: missing_feature (supports DD_HTPP_CLIENT_TAGS_QUERY_STRING instead)
+    Test_Config_ClientIPHeader_Precedence: v2.12.0
+    Test_Config_ClientTagQueryString_Configured: v2.14.0.dev
```
Comment (mabdinur) [inline manifests/python.yml:790]:
```suggestion
    Test_Config_ClientTagQueryString_Configured: v2.15.0
```

File: tests/parametric/test_config_consistency.py [hunk 14]
```diff
@@ -155,11 +153,11 @@ def test_dd_trace_agent_http_url_nonexistent(self, library_env, test_agent, test
...
         # Go tracer reports `<url>/<protocol>/traces` as agent_url in startup log, so use a regex to match the expected suffix
-        if context.library == "golang":
-            match = re.match(r"^http://random-host:9999/.*", resp["dd_trace_agent_url"])
-            assert match is not None, "trace_agent_url does not match expected pattern"
-        else:
-            assert resp["dd_trace_agent_url"] == "http://random-host:9999/"
+        # Ruby tracer reports timeout query string in the agent_url
+        # Python tracer reports only the host and port in the agent_url
+        assert re.match(
+            r"^http://random-host:9999[?|/].*", resp["dd_trace_agent_url"]
```
Comment (mabdinur) [inline tests/parametric/test_config_consistency.py:159]: Use host and port

File: utils/build/docker/ruby/parametric/server.rb [hunk 15]
```diff
@@ -591,13 +593,13 @@ def handle_trace_span_start(req, res)
...
     span = Datadog::Tracing.trace(
-      args.name,
-      service: args.service,
-      resource: args.resource,
-      type: args.type,
-      continue_from: digest
+      start_span_args.name,
+      service: start_span_args.service.empty? ? nil : start_span_args.service,
+      resource: start_span_args.resource.empty? ? nil : start_span_args.resource,
+      type: start_span_args.type.empty? ? nil : start_span_args.type,
```
Comment (mabdinur) [inline utils/build/docker/ruby/parametric/server.rb:601]:
```suggestion
      args.name,
      service: args.service.empty? ? nil : args.service,
      resource: args.resource.empty? ? nil : args.resource,
      type: args.type.empty? ? nil : args.type,
```

**Comment by mabdinur** [comment]
Blocked by: https://github.com/DataDog/dd-trace-rb/pull/4024/files#commit-suggestions. 

Once this PR is merged all ruby tests should be green

File: manifests/python.yml [hunk 16]
```diff
@@ -786,19 +786,23 @@ tests/:
...
-    Test_Config_HttpClientErrorStatuses_FeatureFlagCustom: missing_feature
-    Test_Config_HttpServerErrorStatuses_Default: missing_feature
-    Test_Config_HttpServerErrorStatuses_FeatureFlagCustom: missing_feature
-    Test_Config_IntegrationEnabled_False: missing_feature
-    Test_Config_IntegrationEnabled_True: missing_feature
-    Test_Config_ObfuscationQueryStringRegexp_Configured: missing_feature
-    Test_Config_ObfuscationQueryStringRegexp_Empty: missing_feature
-    Test_Config_UnifiedServiceTagging_CustomService: missing_feature
-    Test_Config_UnifiedServiceTagging_Default: missing_feature
+    Test_Config_HttpClientErrorStatuses_FeatureFlagCustom: missing_feature (xpassing but should fail, this enviornment variable is not supported)
```
Comment (mabdinur) [inline manifests/python.yml:793]:
```suggestion
    Test_Config_HttpClientErrorStatuses_FeatureFlagCustom: missing_feature (xpassing but should fail, this environment variable is not supported)
```

File: manifests/python.yml [hunk 17]
```diff
@@ -786,19 +786,23 @@ tests/:
...
+    Test_Config_HttpServerErrorStatuses_Default: v2.12.0
+    Test_Config_HttpServerErrorStatuses_FeatureFlagCustom: v2.15.0
+    Test_Config_IntegrationEnabled_False:
+      '*': irrelevant (endpoint not implemented)
```
Comment (mabdinur) [inline manifests/python.yml:797]:
```suggestion
      '*': irrelevant (kafka endpoint is not implemented)
```

File: manifests/python.yml [hunk 18]
```diff
@@ -786,19 +786,23 @@ tests/:
...
+      flask-poc: v2.0.0
+    Test_Config_IntegrationEnabled_True:
+      '*': irrelevant (endpoint not implemented)
```
Comment (mabdinur) [inline manifests/python.yml:800]:
```suggestion
      '*': irrelevant (kafka endpoint is not implemented)
```

File: manifests/ruby.yml [hunk 19]
```diff
@@ -267,7 +267,7 @@ tests/:
       test_kafka.py:
         Test_Kafka:
           "*": irrelevant
-          rails70: v0.1 # real version unknown
+          rails70: missing_feature (ruby-kafka integration does not set the kafka topic on spans)
```
Comment (mabdinur) [inline manifests/ruby.yml:270]:
```suggestion
          rails70: missing_feature (ruby-kafka does not set the kafka topic on spans)
```

File: manifests/ruby.yml [hunk 20]
```diff
@@ -408,19 +408,23 @@ tests/:
...
+      '*': irrelevant (endpoint not implemented)
+      rails70: v2.0.0
+    Test_Config_IntegrationEnabled_True:
+      '*': irrelevant (endpoint not implemented)
+      rails70: v2.0.0
     Test_Config_ObfuscationQueryStringRegexp_Configured: missing_feature
-    Test_Config_ObfuscationQueryStringRegexp_Empty: missing_feature
-    Test_Config_UnifiedServiceTagging_CustomService: missing_feature
-    Test_Config_UnifiedServiceTagging_Default: missing_feature
+    Test_Config_ObfuscationQueryStringRegexp_Empty: missing_feature (enviornment variable is not supported)
```
Comment (mabdinur) [inline manifests/ruby.yml:425]:
```suggestion
    Test_Config_ObfuscationQueryStringRegexp_Empty: missing_feature (environment variable is not supported)
```

**Pull request merged**